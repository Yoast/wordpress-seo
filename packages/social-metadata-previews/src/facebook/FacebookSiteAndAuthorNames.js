/* External dependencies */
import React, { Fragment } from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { __, sprintf } from "@wordpress/i18n";

/* Internal dependencies */
import FacebookAuthorName from "./FacebookAuthorName";

const FacebookSiteAndAuthorNamesWrapper = styled.p`
	color: #606770;
	flex-shrink: 0;
	font-size: 12px;
	line-height: 20px;
	overflow: hidden;
	padding: 0;
	text-overflow: ellipsis;
	text-transform: uppercase;
	white-space: nowrap;
	margin: 0;
	position: ${ props => props.mode === "landscape" ? "relative" : "static" };
`;

const FacebookSiteAndAuthorNamesSeparator = styled.span`
	padding-left: 5px;
	padding-right: 5px;
`;

const FacebookSiteUrl = styled.span`
	color: #606770;
	font-size: 12px;
	line-height: 11px;
	text-transform: uppercase;
	overflow: hidden;
`;

/**
 * Renders a FacebookAuthorName component with accompanying elements.
 *
 * @param {string} authorName The author's name.
 *
 * @returns {React.Element} The rendered element.
 */
function renderFacebookAuthorName( authorName ) {
	/* Translators: the context is: SITE | By AUTHOR */
	const by = __( "By", "yoast-components" );

	return (
		<Fragment>
			<FacebookSiteAndAuthorNamesSeparator>|</FacebookSiteAndAuthorNamesSeparator>
			{ by }
			&nbsp;
			<FacebookAuthorName authorName={ authorName } />
		</Fragment>
	);
}

/**
 * Renders a FacebookSiteAndAuthorNames component.
 *
 * @param {object} props The props.
 *
 * @returns {React.Element} The rendered element.
 */
const FacebookSiteAndAuthorNames = ( props ) => {
	const hasAuthorName = props.authorName.length > 0;
	const screenReaderText = hasAuthorName
		/* Translators: 1: site name, 2: post author name */
		? sprintf( __( "%1$s by %2$s", "yoast-components" ), props.siteUrl, props.authorName )
		: props.siteUrl;

	return (
		<Fragment>
			<span className="screen-reader-text">{ screenReaderText }</span>
			<FacebookSiteAndAuthorNamesWrapper aria-hidden="true">
				<FacebookSiteUrl>{ props.siteUrl }</FacebookSiteUrl>
				{ hasAuthorName && renderFacebookAuthorName( props.authorName ) }
			</FacebookSiteAndAuthorNamesWrapper>
		</Fragment>
	);
};

FacebookSiteAndAuthorNames.propTypes = {
	siteUrl: PropTypes.string.isRequired,
	authorName: PropTypes.string,
};

FacebookSiteAndAuthorNames.defaultProps = {
	authorName: "",
};

export default FacebookSiteAndAuthorNames;
