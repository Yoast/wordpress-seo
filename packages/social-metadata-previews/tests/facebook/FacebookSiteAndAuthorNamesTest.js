/* External dependencies */
import React from "react";
import renderer from "react-test-renderer";

/* Internal dependencies */
import FacebookSiteAndAuthorNames from "../../src/facebook/FacebookSiteAndAuthorNames";

describe( "FacebookSiteAndAuthorNames", () => {
	it( "matches the snapshot by default", () => {
		const component = renderer.create(
			<FacebookSiteAndAuthorNames siteUrl="sitename.com" authorName="John Doe" />
		);

		const tree = component.toJSON();
		expect( tree ).toMatchSnapshot();
	} );

	it( "does not contain the author name when no author name was given", () => {
		const component = renderer.create(
			<FacebookSiteAndAuthorNames siteUrl="sitename.com" />
		);

		const tree = component.toJSON();

		/*
		 * Not optimal, but the TestRenderer does not have support for function components. which FacebookAuthorName is.
		 * See: https://reactjs.org/docs/test-renderer.html#testrenderergetinstance
		 */
		const facebookAuthorNameExists = tree[ 1 ].children.some( el => el.type === "span" && el.props.className.startsWith( "FacebookAuthorName" ) );
		expect( facebookAuthorNameExists ).toEqual( false );
	} );
} );
