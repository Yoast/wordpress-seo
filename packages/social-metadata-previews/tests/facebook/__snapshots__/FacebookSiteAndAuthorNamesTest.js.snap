// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FacebookSiteAndAuthorNames matches the snapshot by default 1`] = `
Array [
  <span
    className="screen-reader-text"
  >
    sitename.com by John Doe
  </span>,
  .c3 {
  color: #3b5998;
  font-size: 12px;
  line-height: 11px;
}

.c0 {
  color: #606770;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-size: 12px;
  line-height: 20px;
  overflow: hidden;
  padding: 0;
  text-overflow: ellipsis;
  text-transform: uppercase;
  white-space: nowrap;
  margin: 0;
  position: static;
}

.c2 {
  padding-left: 5px;
  padding-right: 5px;
}

.c1 {
  color: #606770;
  font-size: 12px;
  line-height: 11px;
  text-transform: uppercase;
  overflow: hidden;
}

<p
    aria-hidden="true"
    className="c0"
  >
    <span
      className="c1"
    >
      sitename.com
    </span>
    <span
      className="c2"
    >
      |
    </span>
    By
    Â 
    <span
      className="c3"
    >
      John Doe
    </span>
  </p>,
]
`;
